Disable relocatable shaders with immutable samplers

piotrAMD opened the pull request:
Disable relocatable shaders with immutable samplers as the driver
does not set them up in the descriptor table, so the load with
relocatable offset is not valid.

trenouf reviewed (commented):
LGTM, but I will leave it to Steven to approve.

s-perron commented:
This seems similar to what we were doing in https://github.com/GPUOpen-Drivers/llpc/pull/982.  Why was that check not good enough?

JaxLinAMD commented:
retest this please

piotrAMD commented:
#982 added the check to one leg of the ifdef (LLPC_CLIENT_INTERFACE_MAJOR_VERSION < 41). #1212 handles the "else" case.

I will add a test for that, but it will depend on #1211 being merged first.

piotrAMD commented:
Added test (depends on #1211).

JaxLinAMD commented:
retest this please

amdvlk-admin commented:
### Test summary for commit a61f92527cf7b4902c7563c3e4b2f80c28f0aba1
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 56f2e165b889bd7ca147ff853e6f1372ed90d4fe</li>
<li>SPVGEN: dev 4846370016de9775b7518b51c08260387a7367c4</li>
<li>XGL: dev f67223913cf046cb65adcc3bd3f06c6ad232f150</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 661ac9b7d0d9edcbd07bd1b40c334934ff2297ac</li>
</ul></details>

<details><summary>CTS tests (Failed: 2/222275)</summary><ul><li>Built with version 1.2.5.1</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 30438/55574 (54.8%)</li>
<li>Failed: 0/55574 (0.0%)</li>
<li>Not supported: 25136/55574 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30437/55574 (54.8%)</li>
<li>Failed: 0/55574 (0.0%)</li>
<li>Not supported: 25137/55574 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 30630/55553 (55.1%)</li>
<li><details><summary>Failed: 1/55553 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.synchronization.basic.event.device_set_reset
Stack trace: Script:
synchronizationWrapper->queueSubmit(queue, DE_NULL): VK_ERROR_UNKNOWN at vktSynchronizationBasicEventTests.cpp:117

```
</p></details></li>

<li>Not supported: 24922/55553 (44.9%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34000/55574 (61.2%)</li>
<li><details><summary>Failed: 1/55574 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.synchronization.basic.event.device_set_reset
Stack trace: Script:
synchronizationWrapper->queueSubmit(queue, DE_NULL): VK_ERROR_UNKNOWN at vktSynchronizationBasicEventTests.cpp:117

```
</p></details></li>

<li>Not supported: 21573/55574 (38.8%)</li>
</ul></details>


</ul></details>

kuhar commented on llpc/test/shaderdb/relocatable_shaders/PipelineVsFs_ImmutableSampler.pipe:
```diff
@@ -0,0 +1,77 @@
+
+; BEGIN_SHADERTEST
+; RUN: amdllpc -spvgen-dir=%spvgendir% -enable-relocatable-shader-elf -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST-NOT: Building pipeline with relocatable shader elf
```
This looks like a good candidate for the matching against the new warning: https://github.com/GPUOpen-Drivers/llpc/pull/1210

piotrAMD replied:
Done.

piotrAMD commented:
* Rebased
* Added a nullptr check for innerNode to silence a windows-only build warning in 64-bit optimized build (C28182: Dereferencing NULL pointer).
* Reworked test to use the new warning introduced in #1210.

JaxLinAMD commented:
retest this please
